Content.Shared/GameObjects/Components/Movement/SharedClimbingComponent.cs

```diff
@@ -0,0 +1,66 @@
+﻿using Content.Shared.GameObjects.EntitySystems;
+using Content.Shared.Physics;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameObjects.Components;
+using Robust.Shared.Physics;
+using Robust.Shared.Serialization;
+using System;
+
+namespace Content.Shared.GameObjects.Components.Movement
+{
+    public abstract class SharedClimbingComponent : Component, IActionBlocker, ICollideSpecial
+    {
+        public sealed override string Name => "Climbing";
+        public sealed override uint? NetID => ContentNetIDs.CLIMBING;
+
+        protected ICollidableComponent Body;
+        protected bool IsOnClimbableThisFrame = false;
+
+        protected bool OwnerIsTransitioning
+        {
+            get
+            {
+                if (Body.TryGetController<ClimbController>(out var controller))
+                {
+                    return controller.IsActive;
+                }
+
+                return false;
+            }
+        }
+
+        public abstract bool IsClimbing { get; set; }
+
+        bool IActionBlocker.CanMove() => !OwnerIsTransitioning;
+        bool IActionBlocker.CanChangeDirection() => !OwnerIsTransitioning;
+
+        bool ICollideSpecial.PreventCollide(IPhysBody collided)
+        {
+            if (collided.CollisionLayer == (int) CollisionGroup.VaultImpassable && collided.Entity.HasComponent<IClimbable>())
```

Comment:
CollisionLayer has a collection of flags, meaning that this check would fail whenever the collided body has more collision layers than just vault impassable. This would be fixed by checking if the collided body and the owner share any collision layer flags, excluding vaultimpassable.